File: Micro Serialization Library.JSON (C#)/JsonSerializer.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using MicroSerializationLibrary.Networking;
using MicroSerializationLibrary.Serialization;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Text;

namespace MicroSerializationLibrary.Serialization {
    /// <summary>
    /// The JSON serialization engine is slower than MessagePack, but supports a wider range of objects.
    /// </summary>
    public class JsonSerializer : ISerializationProtocol {

        public byte[] Serialize(object obj) {
            string JsonString = JsonConvert.SerializeObject(new DeserializationWrapper(obj, this));
            return ASCIIEncoding.ASCII.GetBytes(JsonString);
        }

        public object Deserialize(byte[] Data) {
            string JsonString = DeserializationWrapper.ConvertByteToString(Data);
            DeserializationWrapper ObjWrapper = JsonConvert.DeserializeObject<DeserializationWrapper>(JsonString);
            return ObjWrapper.GetInitializedObject(this);
        }

        public object GetPropertyValue(string PropertyName, int PropertyIndex, ref object ObjectData, PropertyReference pr) {
            object v = null;
            Type tt = null;
            object[] indexArray = { 0 };
            Object objD = ObjectData;

            foreach (PropertyReference p in DeserializationWrapper.GetPropertyReferences(pr.Instance)) {
                if (p.Info.Name == pr.Info.Name) {
                    tt = p.Info.GetValue(p.Instance, indexArray).GetType();
                    v = new JObject(objD).ToObject(tt);
                }
            }
            return v;
        }
    }
}

File: Micro Serialization Library.JSON/JsonSerializer.vb
Imports MicroSerializationLibrary.Networking
Imports MicroSerializationLibrary.Serialization
Imports Newtonsoft.Json
Imports Newtonsoft.Json.Linq

''' <summary>
''' The JSON serialization engine is slower than MessagePack, but supports a wider range of objects.
''' </summary>
Public Class JsonSerializer
    Implements ISerializationProtocol

    Public Function Serialize(obj As Object) As Byte() Implements ISerializationProtocol.Serialize
        Dim JsonString As String = JsonConvert.SerializeObject(New DeserializationWrapper(obj, Me))
        Return System.Text.UnicodeEncoding.UTF8.GetBytes(JsonString)
    End Function

    Public Function Deserialize(Data As Byte()) As Object Implements ISerializationProtocol.Deserialize
        Dim JsonString As String = System.Text.UnicodeEncoding.UTF8.GetString(Data)
        Dim ObjWrapper As DeserializationWrapper = JsonConvert.DeserializeObject(Of DeserializationWrapper)(JsonString)
        Return ObjWrapper.GetInitializedObject(Me)
    End Function

    Public Function GetPropertyValue(PropertyName As String, PropertyIndex As Integer, ByRef ObjectData As Object, pr As PropertyReference) As Object Implements ISerializationProtocol.GetPropertyValue
        Try
            Dim v As JValue = DirectCast(ObjectData, JObject)(PropertyName)
            Return v.Value
        Catch ex As Exception
            Dim v As Object
            Dim objD As Object = ObjectData
            DeserializationWrapper.GetPropertyReferences(pr.Instance).ForEach(
                Sub(p)
                    If p.Info.Name = pr.Info.Name Then
                        v = DirectCast(objD, JObject)(PropertyName).ToObject(p.Info.GetValue(p.Instance, {0}).GetType)
                    End If
                End Sub)
            Return v
        End Try
    End Function

End Class

